File: fractran/sim_all.py
#!/usr/bin/env python3
# Simulate all programs in a file for some number of steps and report if they halted.

import argparse
from dataclasses import dataclass
from pathlib import Path
import sys
import time

from base import State, Program
from parse import enum_programs


@dataclass(frozen=True)
class SimResult:
  halted: bool
  num_steps: int
  final_state: State | None


def run(prog: Program, start: State, max_steps: int) -> SimResult:
  state : State | None = start
  num_steps = 0
  while state is not None and num_steps < max_steps:
    prev_state = state
    state, _ = prog.step(state)
    num_steps += 1
  return SimResult(halted=(state is None), num_steps=num_steps, final_state=prev_state)


class OutputWriter:
  def __init__(self, outfile):
    self.outfile = outfile
    self.num_halt = 0
    self.num_total = 0
    self.max_halt_time = 0

  def write(self, prog: Program, result: SimResult) -> None:
    if result.halted:
      status = "HALT"
      self.num_halt += 1
      self.max_halt_time = max(self.max_halt_time, result.num_steps)
    else:
      status = "UNKNOWN"
    self.num_total += 1
    self.outfile.write(f"{prog.fractions_str()}\t{status}\t{result.num_steps}\n")


def main() -> None:
  parser = argparse.ArgumentParser()
  parser.add_argument("infile", type=Path)
  parser.add_argument("outfile", type=Path)
  parser.add_argument("max_steps", type=int)
  parser.add_argument("--print-sec", type=float, default="60")
  args = parser.parse_args()

  with open(args.outfile, "w") as outfile:
    writer = OutputWriter(outfile)
    print_time = time.time() + args.print_sec
    for prog in enum_programs(args.infile):
      start = State.from_int(2, prog.num_registers())
      result = run(prog, start, args.max_steps)
      writer.write(prog, result)
      if time.time() >= print_time:
        print(f"...  Total: {writer.num_total:_d}  Halted: {writer.num_halt:_d} ({writer.num_halt/writer.num_total:.0%})  Max Halt Time: {writer.max_halt_time:_d}  ({time.process_time():_f}s)")
        print_time = time.time() + args.print_sec

  print(f"Finished:  Total: {writer.num_total:_d}  Halted: {writer.num_halt:_d} ({writer.num_halt/writer.num_total:.0%})  Max Halt Time: {writer.max_halt_time:_d}  ({time.process_time():_f}s)")

if __name__ == "__main__":
  main()

File: fractran/rust/src/bin/sim-all.rs
// Simulate all programs in a file for some number of steps and keep track of halting times.

use std::time::{Duration, Instant};

use clap::Parser;
use indicatif::ParallelProgressIterator;
use rayon::prelude::*;

use fractran::parse::{load_lines, parse_program};
use fractran::program::{Int, SimResult, State};

#[derive(Parser, Debug)]
#[command(version, about, long_about = None)]
struct Args {
    filename: String,
    step_limit: Int,
}

struct TaskResult {
    sim: SimResult,
    duration: Duration,
}

// Helper function to run the simulation and collect results
fn parse_and_sim(program_str: &str, step_limit: Int) -> TaskResult {
    let start_time = Instant::now();
    let prog = parse_program(program_str);
    let mut state = State::start(&prog);
    let sim_result = prog.run(&mut state, step_limit);

    if sim_result.halted {
        println!("  Halt: {} steps: {}", sim_result.total_steps, program_str);
    }

    TaskResult {
        sim: sim_result,
        duration: start_time.elapsed(),
    }
}

fn main() {
    let args = Args::parse();

    // 2. Load all program strings
    let programs = load_lines(&args.filename);
    let num_programs = programs.len();

    println!(
        "Simulating {} programs for {} steps using {} threads",
        num_programs,
        args.step_limit,
        rayon::current_num_threads()
    );
    let wallclock_start_time = Instant::now();

    // 3. Parallel Execution using Rayon
    let results: Vec<TaskResult> = programs
        .par_iter()
        .progress_count(num_programs as u64)
        .filter_map(|program_str| Some(parse_and_sim(program_str, args.step_limit)))
        .collect(); // Collect results back into a Vec on the main thread

    let wallclock_time_sec = wallclock_start_time.elapsed().as_secs_f64();

    // 4. Summarize results
    let halted_count = results.iter().filter(|r| r.sim.halted).count();
    let frac_halted = halted_count as f64 / num_programs as f64;
    let max_halt_steps = results
        .iter()
        .map(|r| if r.sim.halted { r.sim.total_steps } else { 0 })
        .max()
        .unwrap();
    let total_steps_simulated: Int = results.iter().map(|r| r.sim.total_steps).sum();

    let wallclock_rate = total_steps_simulated as f64 / wallclock_time_sec;

    let max_runtime_sec = results
        .iter()
        .map(|r| r.duration)
        .max()
        .unwrap()
        .as_secs_f64();
    let total_thread_runtime_sec = results
        .iter()
        .map(|r| r.duration)
        .sum::<Duration>()
        .as_secs_f64();
    let mean_runtime_sec = total_thread_runtime_sec / num_programs as f64;

    println!("\nResults Summary:");
    println!(
        "  Halted: {} / {}  ({:.2}%)",
        halted_count,
        num_programs,
        frac_halted * 100.0
    );
    println!("  Max Halt Steps: {}", max_halt_steps);
    println!("  Total steps: {}", total_steps_simulated);
    println!("Wallclock Time:");
    println!("  {:.2} seconds", wallclock_time_sec);
    println!("  {:.2} Million steps/s", wallclock_rate / 1_000_000.0);
    println!("CPU Time:");
    println!("  Total: {:.2} core-sec", total_thread_runtime_sec);
    println!("  Mean:  {:.2} core-sec / program", mean_runtime_sec);
    println!("  Max:   {:.2} core-sec", max_runtime_sec);
}
